core/src/Cuda/Kokkos_Cuda_abort.hpp

```diff
@@ -56,21 +56,20 @@ extern "C" {
 /*  Cuda runtime function, declared in <crt/device_runtime.h>
  *  Requires capability 2.x or better.
  */
-extern __device__ void __assertfail(const void *message, const void *file,
-                                    unsigned int line, const void *function,
-                                    size_t charsize);
+[[noreturn]] __device__ void __assertfail(const void *message, const void *file,
```

Comment:
I still see
```
extern __host__ __device__ void __assertfail(const char * __assertion,
                                             const char *__file,
                                             unsigned int __line,
                                             const char *__function,
                                             size_t charsize);
```
in `crt/common_functions.h`. Adding `[[noreturn]]` here sounds dangerous.

Reply:
I think __assertfail is necessarily noreturn, right? Or is that not true? [[noreturn]] shouldn't change the ABI of the routine...

Reply:
Oh, is the concern that we're changing the definition for the routine? Hm... Maybe I'll just add an infinite loop after the __assertfail call instead...

Reply:
https://en.cppreference.com/w/cpp/language/attributes/noreturn says:
```
If a function is declared with [[noreturn]] in one translation unit, and the same function is declared without [[noreturn]] in another translation unit, the program is ill-formed; no diagnostic required.
```
That's my only concern.

Reply:
Yeah, I think you're right.

Reply:
I assume the infinite loop suggestion was a joke right?

Reply:
Well, no, but maybe you can help illuminate any potential issue. `__assertfail` is semantically `[[noreturn]]`, but it's not marked as such. You need to prove to the compiler that a function actually is no return, so putting an infinite loop after the `__assertfail` would do that. However, this loop would never be entered. Could it still cause issues?